File: 2025-12/src/day_16/boj_10868/Main.java
package day_16.boj_10868;

import java.io.*;
import java.util.*;

public class Main {

    static final int MAX = 1_000_000_000;

    static int[] arr;
    static int[] tree;

    static void init(int node, int start, int end) {
        if (start == end) {
            tree[node] = arr[start];
            return;
        }

        int mid = (start + end) / 2;

        init(node * 2, start, mid);
        init(node * 2 + 1, mid + 1, end);
        tree[node] = Math.min(tree[node * 2], tree[node * 2 + 1]);
    }

    static int queryMin(int node, int start, int end, int left, int right) {
        if (left > end || right < start) return MAX;
        if (left <= start && end <= right) return tree[node];

        int mid = (start + end) / 2;

        int leftMin = queryMin(node * 2, start, mid, left, right);
        int rightMin = queryMin(node * 2 + 1, mid + 1, end, left, right);
        return Math.min(leftMin, rightMin);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        arr = new int[1 + n];
        for (int i = 1; i <= n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }

        tree = new int[4 * n];
        init(1, 1, n);

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            sb.append(queryMin(1, 1, n, a, b)).append("\n");
        }

        System.out.println(sb);
    }
}

File: 2025-12/src/day_16/boj_10868/Main.cpp
#include <bits/stdc++.h>
using namespace std;

constexpr int MAX = 1000000000;
int arr[100001];
int tree[4 * 100000];

void init(int node, int start, int end) {
    if (start == end) {
        tree[node] = arr[start];
        return;
    }

    int mid = (start + end) / 2;

    init(node * 2, start, mid);
    init(node * 2 + 1, mid + 1, end);
    tree[node] = min(tree[node * 2], tree[node * 2 + 1]);
}

int queryMin(int node, int start, int end, int left, int right) {
    if (left > end || right < start) return MAX;
    if (left <= start && end <= right) return tree[node];

    int mid = (start + end) / 2;

    int leftMin = queryMin(node * 2, start, mid, left, right);
    int rightMin = queryMin(node * 2 + 1, mid + 1, end, left, right);
    return min(leftMin, rightMin);
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    for (int i = 1; i <= n; i++) {
        cin >> arr[i];
    }

    init(1, 1, n);

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        cout << queryMin(1, 1, n, a, b) << '\n';
    }
}
